[Auto Parallel]: Support std::vector<phi::Tensor> input and output for DistTensor.

GhostScreaming opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
Others

### PR changes
Others

### Description
Pcard-73145

Support `std::vector<phi::Tensor>` input and output for `DistTensor`. Meanwhile `no_need_buffer` type is also supported. `Concat`, `Broadcast_Tensors`, `Unbind` forward and backward are verified. Following operators need to be supported later (along with their backward op if exists): `check_finite_and_unscale`, `coalesce_tensor`, `meshgrid`, `update_loss_scaling`, `einsum`.

Foward operators have output of `std::tuple<...>` are not supported now.

GhostScreaming committed:
[WIP] Support std::vector<phi::Tensor> input and output for DistTensor.
Concat forward and backward are verified.

chenwhql commented on paddle/fluid/eager/tensor_wrapper.h:
```diff
@@ -66,6 +70,17 @@ class TensorWrapper {
+        // TODO(jiabin): It's not a good idea to set memory size to zero, find
+        // another way and change this.
+        intermidiate_tensor_.set_impl(std::make_shared<phi::distributed::DistTensor>(
+            std::make_shared<phi::Allocation>(nullptr, 0, tensor.place()),
```
这个构造函数已经移除了，需要看下最新的构造方式

GhostScreaming committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into dist_tensor_support

GhostScreaming committed:
Polish code for new dist tensor implementation.

GhostScreaming committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into dist_tensor_support

GhostScreaming committed:
Fix bug of DistTensor upgrade. Add support functions for std::vector<Tensor> -> std::vector<Tensor>.

GhostScreaming committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into dist_tensor_support

GhostScreaming committed:
Add support for DistTensor type of std::vector<phi::Tensor> as input or output of operators.
Following testcases are passed.
1. concat: std::vector<phi::Tensor> -> phi::Tensor
2. unbind: phi::Tensor -> std::vector<phi::Tensor>
3. broadcast_tensors: std::vector<phi::Tensor> -> std::vector<phi::Tensor>

GhostScreaming committed:
Polish code. Remove useless comments.

GhostScreaming committed:
Add update_loss_scaling in skip_op_lists.

chenwhql commented on paddle/phi/api/yaml/generator/dist_api_gen.py:
```diff
@@ -71,10 +71,31 @@
+SINGLE_OUT_VECTOR_CREATION_TEMPLATE = """
+    auto dist_out = SetKernelDistOutput({}, &api_output);
+    std::vector<phi::DenseTensor*> dense_out(dist_out.size());
+    for (size_t i=0; i<dist_out.size(); i++) {{
```
一般起始条件和循环条件代码之间会有空格：
`for (size_t i = 0; i < dist_out.size(); i++)`

GhostScreaming replied:
Done, thx

chenwhql commented on paddle/fluid/eager/tensor_wrapper.h:
```diff
@@ -66,6 +68,16 @@ class TensorWrapper {
+        phi::distributed::DistTensor* dist_tensor =
+            static_cast<phi::distributed::DistTensor*>(tensor.impl().get());
+        intermidiate_tensor_.set_impl(
+            std::make_shared<phi::distributed::DistTensor>(
```
这里新构造tensor会触发重切分，我们还需要讨论一下，这样能过的话，也可以先这样写

GhostScreaming replied:
目前在concat上验证，是没有问题的。可以先加一个TODO？

chenwhql replied:
没关系，后面我来处理下

chenwhql commented on paddle/phi/api/yaml/generator/dist_api_gen.py:
```diff
@@ -71,10 +71,31 @@
+        dense_out[i] = const_cast<phi::DenseTensor*>(&dist_out[i]->value());
+    }}
+"""
+VECTOR_SINGLE_OUT_CREATION_TEMPLATE = """
```
VECTOR_SINGLE_OUT和上面的SINGLE_OUT_VECTOR有什么区别

GhostScreaming replied:
这个代码是废弃的，已删除

chenwhql commented on paddle/phi/api/yaml/generator/dist_api_gen.py:
```diff
@@ -91,12 +112,31 @@
 # TODO(chenweihang): InferSPMD function design
 SINGLE_DIST_META_IN_TEMPLATE = """MakeMetaTensor(*{}.impl()), """
 # TODO(chenweihang): support vector and optional args later
```
已经支持了话，可以帮忙把这个TODO更新下

GhostScreaming replied:
Done, thx

chenwhql commented on paddle/phi/api/yaml/generator/dist_api_gen.py:
```diff
@@ -175,6 +235,14 @@
 #     types : [], list of output types
 #     out_size_expr : [], expression for getting size of vector<Tensor>
 
+skip_op_lists = [
```
这里后续还会处理吗？是否需要加个TODO说明下

GhostScreaming replied:
后续会支持std::tuple<...>类型的输入输出，已更新

chenwhql commented on paddle/phi/api/yaml/generator/dist_api_gen.py:
```diff
@@ -335,13 +412,24 @@ def generate_infer_spmd_code(self) -> str:
             if infer_meta['param'] is not None
             else input_names + attr_names
         )
+        input_mega_code = ""
```
mega -> meta？

GhostScreaming replied:
Done, thx

chenwhql commented on paddle/phi/api/yaml/generator/dist_api_gen.py:
```diff
@@ -387,7 +486,7 @@ def generate_reshard_input_code(self) -> str:
         return INPUT_RESHARD_TEMPLATE.format()
 
     # override BaseAPI's method
```
这个注释应该有点问题，帮忙删掉吧

GhostScreaming replied:
Done, thx

chenwhql commented on paddle/phi/api/yaml/generator/dist_api_gen.py:
```diff
@@ -410,6 +509,27 @@ def generate_dense_input(
 
         return input_tensor_code
 
+    # override BaseAPI's method
```
这里也需要删掉，这应该是子类独有的方法了

GhostScreaming replied:
Done, thx

chenwhql commented on paddle/phi/api/yaml/generator/dist_bw_api_gen.py:
```diff
@@ -27,6 +27,13 @@
     auto dist_out = SetKernelDistOutput({});
     auto dense_out = const_cast<phi::DenseTensor*>(&dist_out->value());
 """
+PARALLEL_OUT_CREATION_TEMPLATE = """
```
是不是用VECTOR更合适一些，SINGLE这里也是给Parallel用的

GhostScreaming replied:
Done, thx

GhostScreaming committed:
Polish code.

LiYuRio commented on paddle/phi/api/lib/data_transform.cc:
```diff
@@ -632,5 +632,47 @@ std::shared_ptr<phi::distributed::DistTensor> PrepareDataForDistTensor(
+                         const TransformFlag& transform_flag,
+                         bool is_stride_kernel) {
+  std::vector<std::shared_ptr<phi::distributed::DistTensor>> out;
+  for (auto x : input) {
```
这里可以改成auto&，减少拷贝

LiYuRio commented on paddle/phi/api/lib/data_transform.cc:
```diff
@@ -632,5 +632,47 @@ std::shared_ptr<phi::distributed::DistTensor> PrepareDataForDistTensor(
+                                     dense_tensor.meta().is_contiguous()))) {
+        out.push_back(
+            std::static_pointer_cast<phi::distributed::DistTensor>(tensor_in));
+        continue;
```
理论上这里可以写到else分支里，不用continue

LiYuRio reviewed (approved):
小问题可以下个pr改
